### PR TITLE
Node ctor validation

### LINKED ISSUE #261: Node constructor validation
`BlankNode`, `Variable` and `IRI` constructors should validate the passed in values.

Additionally, add `BlankNode::make_unchecked` and friends to be able to construct nodes without these checks (probably still add asserts in those tho).

### PR DESCRIPTION
**Pull request by mcb5637**
close #261

### CODE REVIEW

File: src/rdf4cpp/util/CharMatcher.hpp [hunk 1]
```diff
@@ -1,27 +1,91 @@
...
+};
 
 template<typename T>
-concept CharMatcher = requires(const T a, int c) {{a.match(c)} -> std::same_as<bool>; };
+concept CharMatcher = requires(T const a, int c) {
+    {
+        a.match(c)
+    } -> std::same_as<bool>;
+    {
+        T::SIMD_RANGE_NUM
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.hpp]: wrong constant style

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+        if (HWY_UNLIKELY(idx == count))
+            return;
+
+        size_t const remaining = count - idx;
+        HWY_DASSERT(0 != remaining && remaining < N);
+        hwy::HWY_NAMESPACE::Vec<D> const v = LoadNOr(no, d, in + idx, remaining);
+        func(d, v);
+    }
+
+    std::optional<bool> try_match_simd_impl(std::string_view data, std::array<CharRange, 3> const &ranges, std::string_view single) {
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp]: This function needs more documentation what the idea of operation is

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        // 5 - logic
+        // ? - singles
+        // 16 overall
+        bool found_unicode = false;
+        bool r = true;
+        using D = hwy::HWY_NAMESPACE::ScalableTag<int8_t>;  //NOLINT
+        using V = hwy::HWY_NAMESPACE::VFromD<D>;             //NOLINT
+        D d;
+        bool r0a = ranges[0].first < ranges[0].last;
+        V zero = r0a ? hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(ranges[0].first)) : hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(single.at(0)));
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp]: Is it possible to alias the `hwy::HWY_NAMESPACE`? This adds a lot of noise

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        D d;
+        V zero = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(0));
+        V m0 = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(match[0]));
+        V m1 = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(match[1]));
+        V m2 = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(match[2]));
+        V m3 = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(match[3]));
+
+        Foreach(d, reinterpret_cast<int8_t const *>(data.data()), data.size(), zero, [&](auto d, auto in_vec) HWY_ATTR {
+            auto m = in_vec == m0;
+            m = hwy::HWY_NAMESPACE::Or(m, in_vec == m1);
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp]: documentation here as well

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        // register usage:
+        // 1 - data
+        // 6 - ranges
+        // 5 - logic
+        // ? - singles
+        // 16 overall
+        bool found_unicode = false;
+        bool r = true;
+        using D = hwy::HWY_NAMESPACE::ScalableTag<int8_t>;  //NOLINT
+        using V = hwy::HWY_NAMESPACE::VFromD<D>;             //NOLINT
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp]: A very short comment what D and V are would be helpful, I know they say what they are in the documentation but having it here would be helpful

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        size_t si_num = single.size();
+        if (si_num > si.size()) {
+            assert(false);
+            return std::nullopt;
+        }
+        for (size_t i = 0; i < si_num; ++i) {
+            si[i] = hwy::HWY_NAMESPACE::Set(d, static_cast<int8_t>(single[i]));
+        }
+
+        Foreach(d, reinterpret_cast<int8_t const *>(data.data()), data.size(), zero, [&](auto d, auto in_vec) HWY_ATTR {
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp:81]: Is foreach just a badly named `fold`? I'm confused what the `zero` param does

**Review by liss-h — CHANGES_REQUESTED** [review]
Thx for the PR.

I think the SIMD stuff needs a lot more documentation.
At least to me the library makes understanding the SIMD stuff harder than the raw C instruction equivalents, but it is what it is. At least we get dispatch.

So I hope that adding some comments and reducing noise the library introduces helps. I will have a deeper look then

File: src/rdf4cpp/BlankNode.cpp [hunk 7]
```diff
@@ -1,14 +1,16 @@
...
 #include <rdf4cpp/writer/TryWrite.hpp>
+#include <uni_algo/all.h>
 
 namespace rdf4cpp {
 BlankNode::BlankNode() noexcept : Node{storage::identifier::NodeBackendHandle{{}, storage::identifier::RDFNodeType::BNode, {}}} {
 }
 
 BlankNode::BlankNode(std::string_view identifier, storage::DynNodeStoragePtr node_storage)
-    : Node{storage::identifier::NodeBackendHandle{node_storage.find_or_make_id(storage::view::BNodeBackendView{.identifier = identifier, .scope = std::nullopt}),
+    : Node{storage::identifier::NodeBackendHandle{node_storage.find_or_make_id(storage::view::BNodeBackendView{.identifier = validate_bnode_name(identifier), .scope = std::nullopt}),
```
Comment (liss-h) [inline src/rdf4cpp/BlankNode.cpp]: this could use `make_unchecked`

File: src/rdf4cpp/util/CharMatcher.cpp (hunk 6 above)
Comment (mcb5637) [reply src/rdf4cpp/util/CharMatcher.cpp:81]: zero is the placeholder that gets used, if there is no full vector of data left

**Comment by mcb5637** [comment]
in preparation for #267 , also added ParsingError and throw it instead of runtime_error

File: src/rdf4cpp/IRI.hpp [hunk 8]
```diff
@@ -36,6 +38,7 @@ struct IRI : Node {
      * Constructs an IRI object from a IRI string
      * @param iri IRI string
      * @param node_storage optional custom node_storage used to store the IRI
+     * @throw std::invalid_argument if iri is invalid
```
Comment (liss-h) [inline src/rdf4cpp/IRI.hpp]: needs to be updated to `ParsingError`

File: src/rdf4cpp/IRI.hpp [hunk 9]
```diff
@@ -45,12 +48,20 @@ struct IRI : Node {
     [[nodiscard]] static IRI make_null() noexcept;
 
     /**
-     * Constructs an IRI object from a IRI string
+     * Constructs an IRI object from a IRI string.
      * @param iri IRI string
      * @param node_storage optional custom node_storage used to store the IRI
+     * @throw std::invalid_argument if iri is invalid
```
Comment (liss-h) [inline src/rdf4cpp/IRI.hpp]: same here

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+#include <stdexcept>
+
+
+HWY_BEFORE_NAMESPACE();  // at file scope
+namespace rdf4cpp::util::char_matcher_detail::HWY_NAMESPACE {
+    // Calls `func(d, v)` for each input vector; out of bound lanes with index i >=
+    // `count` are instead taken from `no[i % Lanes(d)]`.
+    // if func returns false, exits early
+    // from hwy/contrib/algo/transform-inl.h
+    template<class D, class Func, typename T = hwy::HWY_NAMESPACE::TFromD<D>>
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp]: typename

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+#include <hwy/highway.h>
+#include <stdexcept>
+
+
+HWY_BEFORE_NAMESPACE();  // at file scope
+namespace rdf4cpp::util::char_matcher_detail::HWY_NAMESPACE {
+    // Calls `func(d, v)` for each input vector; out of bound lanes with index i >=
+    // `count` are instead taken from `no[i % Lanes(d)]`.
+    // if func returns false, exits early
+    // from hwy/contrib/algo/transform-inl.h
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp]: Why did you copy the code here? Is this a seperate library?

File: src/rdf4cpp/util/CharMatcher.hpp [hunk 12]
```diff
@@ -1,27 +1,105 @@
...
+    explicit constexpr ASCIIPatternMatcher(char const (&str)[n]) noexcept
+        : pattern(str) {
+    }
 
     [[nodiscard]] constexpr bool match(int c) const noexcept {
-        auto ch = static_cast<char>(c);
-        if (c != static_cast<int>(ch))  // not asciii
-            return false;
-        return pattern.find(ch) != std::string_view::npos;
+    auto ch = static_cast<char>(c);
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.hpp]: indentation

File: src/rdf4cpp/util/CharMatcher.hpp [hunk 13]
```diff
@@ -1,27 +1,105 @@
...
+ * @return
+ */
+bool contains_any(std::string_view data, std::array<char, 4> match);
+
+/**
+ * for test cases only!
+ * iterates over each available highway target and calls func specifically with only that target selected.
+ * @param func
+ */
+void test_simd_foreach_supported(void(*func)(std::string_view target_name));
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.hpp]: Put this into the test cases if its only used there

File: src/rdf4cpp/util/CharMatcher.hpp [hunk 14]
```diff
@@ -1,27 +1,105 @@
...
+    char first = '\0';
+    char last = '\0';
+};
 
 template<typename T>
-concept CharMatcher = requires(const T a, int c) {{a.match(c)} -> std::same_as<bool>; };
+concept CharMatcher = requires(T const a, int c) {
+    {
+        a.match(c)
+    } -> std::same_as<bool>;
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.hpp]: `convertible_to`

File: src/rdf4cpp/util/CharMatcher.hpp [hunk 15]
```diff
@@ -1,27 +1,105 @@
...
+    {
+        T::simd_range_num
+    } -> std::same_as<const size_t &>;
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.hpp]: `convertible_to<size_t>`

File: src/rdf4cpp/util/CharMatcher.hpp [hunk 16]
```diff
@@ -1,27 +1,105 @@
...
+    {
+        T::fail_if_unicode
+    } -> std::same_as<const bool &>;
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.hpp]: `convertible_to<bool>`

File: src/rdf4cpp/util/CharMatcher.hpp [hunk 17]
```diff
@@ -1,27 +1,105 @@
...
+    {
+        a.simd_ranges()
+    } -> std::same_as<std::array<CharRange, T::simd_range_num>>;
+    {
+        a.simd_singles()
+    };  // -> std::same_as<std::string_view>;
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.hpp]: Why is this commented out?

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+        return try_match_simd_impl(data, ranges, single);
+    }
+    std::optional<bool> try_match_simd_impl_3_21(std::string_view data, std::array<CharRange, 3> const &ranges, datatypes::registry::util::ConstexprString<21> const &single) {
+        return try_match_simd_impl(data, ranges, single);
+    }
+    std::optional<bool> try_match_simd_impl_1_1(std::string_view data, std::array<CharRange, 1> const &ranges, datatypes::registry::util::ConstexprString<1> const &single) {
+        return try_match_simd_impl(data, ranges, single);
+    }
+
+    bool contains_any_impl(std::string_view data, std::array<char, 4> match) {
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp]: why is `match` not simply `string_view`?

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+        bool r = false;
+
+        using D = ScalableTag<int8_t>;  //NOLINT  tag type, selects the used vector type
+        using V = Vec<D>;             //NOLINT vector type
+        D const d;
+
+        // elements from this vector are used, if data is not a multiple of Lanes(d)
+        // should not influence final result
+        // => comparisons need to evaluate as false for the logic to work
+        V const zero = Set(d, static_cast<int8_t>(0));
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp:139]: So, we need to `assert` that `match` does not contain `\0`

File: src/rdf4cpp/util/CharMatcher.cpp [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+
+        Foreach(d, reinterpret_cast<int8_t const *>(data.data()), data.size(), zero, [&](auto d, auto in_vec) HWY_ATTR {
+            // if bit 7 is set, the char belongs to a unicode sequence
+            // => std::nullopt
+            if (!AllTrue(d, HighestSetBitIndex(in_vec) < unicode_bit_index)) {
+                found_unicode = true;
+                return false;
+            }
+
+            // highway doc: on x86 < and > are 1 instruction for signed ints (3 for unsigned)
```
Comment (liss-h) [inline src/rdf4cpp/util/CharMatcher.cpp]: this comment would be more useful higher up where the ranges are set-up

**Review by liss-h — CHANGES_REQUESTED** [review]
Thanks for the changes, much more readable now. I've left some additional comments.

The thing with the templates is really ugly, but I couldn't find a way to make it work with the export macro either. I'm pondering if it makes sense to use a `static thread_local std::vector` for the required buffers (or maybe `flex_array`) to avoid the template all together, but idk

File: src/rdf4cpp/util/CharMatcher.cpp (hunk 11 above)
Comment (mcb5637) [reply src/rdf4cpp/util/CharMatcher.cpp]: the highway one does not allow for early return. added a comment explaining it.

File: src/rdf4cpp/util/CharMatcher.cpp (hunk 18 above)
Comment (mcb5637) [reply src/rdf4cpp/util/CharMatcher.cpp]: because string_view is arbitrary length.
and the code is more efficient, if i know the length at compile time.

**Review by liss-h — APPROVED** [review]

**Pull request merged**